lib/Monads/NonDetMonad.thy

```diff
@@ -722,6 +722,38 @@ where
 notation (output)
   whileLoopE  ("(whileLoopE (_)//  (_))" [1000, 1000] 1000)
 
+
+section "More Monadic Combinators"
```

Comment:
I agree with the Misc/More/Extra concern (the name is probably my fault), but I do think they should go into their own section, because there is a foundational difference between these combinators and the ones we had before.

For all of these new ones, conditions can themselves be monads (which is why they are `ifM` etc). We have explicitly avoided that so far, because it is annoying to reason about constructs where tests can have side effects. In `MCS` we finally hit loops and conditions in C where there was no good way to avoid that any more, so I gave up on my resistance to adding them to the library, but I think it's worth pointing out that these are different to the other ones.

So the question is what is a good section title? Maybe `Combinators that have conditions with side effects`?

Reply:
Ah, yes, that is what comments/titles should be for. Without deep examination I didn't figure out they should be a special class. Thumbs up.